### ISSUE TITLE
No matching entry for selector parameter with value ''

### ISSUE THREAD

**Issue by walterheck** [elastic/puppet-elasticsearch#503]
Hi, 

I'm getting the above error on a centos 7 box with puppet 4.2.1. Here's what I see:

```
[root@elasticsearch ~]# puppet agent -t --environment walter_fixes_shreejitbranch
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: No matching entry for selector parameter with value '' at /etc/puppetlabs/code/environments/walter_fixes_shreejitbranch/modules/elasticsearch/manifests/instance.pp:102:21 on node elasticsearch
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
```

these are the offending lines from that file:

```
 102    $notify_service = $elasticsearch::restart_on_change ? {
 103      true  => Elasticsearch::Service[$name],
 104      false => undef,
 105    }
```

which means that $elasticsearch::restart_on_change is actually an empty string
    but in the class definition of elasticsearch class it says:

```
   201  class elasticsearch(
   202    $ensure                = $elasticsearch::params::ensure,
   203    $status                = $elasticsearch::params::status,
   204    $restart_on_change     = $elasticsearch::params::restart_on_change,
   205    $autoupgrade           = $elasticsearch::params::autoupgrade,
   206    $version               = false,
   207    $package_provider      = 'package',
   208    $package_url           = undef,
   209    $package_dir           = $elasticsearch::params::package_dir,
   210    $package_name          = $elasticsearch::params::package,
   211    $package_pin           = true,
   212    $purge_package_dir     = $elasticsearch::params::purge_package_dir,
   213    $package_dl_timeout    = $elasticsearch::params::package_dl_timeout,
   214    $proxy_url             = undef,
   215    $elasticsearch_user    = $elasticsearch::params::elasticsearch_user,
   216    $elasticsearch_group   = $elasticsearch::params::elasticsearch_group,
   217    $configdir             = $elasticsearch::params::configdir,
   218    $purge_configdir       = $elasticsearch::params::purge_configdir,
   219    $service_provider      = 'init',
   220    $init_defaults         = undef,
   221    $init_defaults_file    = undef,
   222    $init_template         = undef,
   223    $config                = undef,
   224    $datadir               = $elasticsearch::params::datadir,
   225    $plugindir             = $elasticsearch::params::plugindir,
   226    $plugintool            = $elasticsearch::params::plugintool,
   227    $java_install          = false,
   228    $java_package          = undef,
   229    $manage_repo           = false,
   230    $repo_version          = undef,
   231    $logging_file          = undef,
   232    $logging_config        = undef,
   233    $logging_template      = undef,
   234    $default_logging_level = $elasticsearch::params::default_logging_level,
   235    $repo_stage            = false,
   236    $instances             = undef,
   237    $instances_hiera_merge = false,
   238    $plugins               = undef,
   239    $plugins_hiera_merge   = false
   240  ) inherits elasticsearch::params {
```

    so it defaults to whatever is in the params class
    in the params class then, it says:

```    
    30  class elasticsearch::params {
    31
    32    #### Default values for the parameters of the main module class, init.pp
    33
    34    # ensure
    35    $ensure = 'present'
    36
    37    # autoupgrade
    38    $autoupgrade = false
    39
    40    # service status
    41    $status = 'enabled'
    42
    43    # restart on configuration change?
    44    $restart_on_change = true
    45
    46    # Purge configuration directory
    47    $purge_configdir = false
    48
    49    $purge_package_dir = false
```

the profile I have is like this:

```
class profile::elasticsearch{  
  # resources required for elasticsearch
  include ::java
  include ::elasticsearch
  
  # load elasticsearch instance details fore hiera
  $instances = hiera_hash('profile::elasticsearch::instances', undef)
  
  if $instances {
    create_resources('elasticsearch::instance', $instances)
  } else {
    fail("No elasticsearch instances found in hiera for class profile::elasticsearch on ${::fqdn}.")
  }
  
}
```

and my hiera data looks like this:

```
---
elasticsearch::manage_repo: true
elasticsearch::repo_version: '2.x'
profile::elasticsearch::instances:
  elasticsearch:
    ensure: present
    status: enabled
    config:
      network.host: "%{::ipaddress_eth0}"
      http.port: '9200'
profile::kibana::elasticsearch_url: 'http://elasticsearch:9200'
profile::kibana::manage_user: 'true'
profile::kibana::kibana4_user: 'kibana4'
profile::kibana::kibana4_group: 'kibana4'
profile::kibana::kibana4_gid: '200'
profile::kibana::kibana4_uid: '200'
profile::kibana::package_ensure: '4.2.1-linux-x64'
profile::kibana::repo_version: '4.2'
profile::kibana::install_dir: '/etc/kibana'
profile::kibana::symlink: 'false'
profile::kibana::src_target: '/usr/src/kibana'
profile::kibana::package_provider: 'archive'
profile::kibana::package_download_url: 'https://download.elastic.co/kibana/kibana/kibana-4.2.1-linux-x64.tar.gz'
profile::kibana::port: '5601'
```

I don't see what the problem is, anyone else can reproduce?

**Comment by walterheck** [comment]
I noticed this module is actually employing questionable code patterns: the elasticsearch::instance define is using parameters from the main elasticsearch class, which it can't and shouldn't rely on already being loaded as it can't guarantee that.

**Comment by electrical** [comment]
If you would change the `include ::elasticsearch` to `class { '::elasticsearch': }` it should work just fine.

For the 'questionable code patterns' its very common to pull variables from the main class in defines so you don't have to define the same variables multiple times everywhere.

**Comment by walterheck** [comment]
> If you would change the include ::elasticsearch to class { '::elasticsearch': } it should work just fine.

huh? why does that make a difference? I'm a puppetlabs trainer and I know there are subtle differences between the two syntaxes (like include can be used multiple times vs class{} only once, etc. I have never encountered this though?

> For the 'questionable code patterns' its very common to pull variables from the main class in defines so you don't have to define the same variables multiple times everywhere.
something being very common doesn't make it proper. In my opinion it is okay for a class to access out of scope variables but for a define I find it better to send the variables it needs to access as parameters. This issue being my case in point :)

**Comment by electrical** [comment]
Even if it was a variable in the define it self you would still run into the same issue.

The include can indeed be done multiple times but doesn't initialize the class it self which has a function on its own like ensuring the sequencing of different defines.
Include is only useful if you want to have access to the variables it has, but in this case you want to initialize the class.

**Comment by walterheck** [comment]
Thanks for the help, working now.
